File: bridges/nodejs/src/sdk/memory.ts
import path from 'node:path'
import fs from 'node:fs'

import { SKILL_PATH, SKILLS_PATH } from '@bridge/constants'

interface MemoryOptions<T> {
  name: string
  defaultMemory?: T
}

export class Memory<T = unknown> {
  private readonly memoryPath: string
  private readonly name: string
  private readonly defaultMemory: T | undefined
  private isFromAnotherSkill: boolean

  constructor(options: MemoryOptions<T>) {
    const { name, defaultMemory } = options

    this.name = name
    this.defaultMemory = defaultMemory
    this.memoryPath = path.join(SKILL_PATH, 'memory', `${this.name}.json`)
    this.isFromAnotherSkill = false

    if (this.name.includes(':') && this.name.split(':').length === 3) {
      this.isFromAnotherSkill = true

      const [domainName, skillName, memoryName] = this.name.split(':')
      this.memoryPath = path.join(
        SKILLS_PATH,
        domainName as string,
        skillName as string,
        'memory',
        `${memoryName}.json`
      )
    }
  }

  /**
   * Clear the memory and set it to the default memory value
   * @example clear()
   */
  public async clear(): Promise<void> {
    if (!this.isFromAnotherSkill) {
      await this.write(this.defaultMemory as T)
    } else {
      throw new Error(
        `You cannot clear the memory "${this.name}" as it belongs to another skill`
      )
    }
  }

  /**
   * Read the memory
   * @example read()
   */
  public async read(): Promise<T> {
    if (this.isFromAnotherSkill && !fs.existsSync(this.memoryPath)) {
      throw new Error(
        `You cannot read the memory "${this.name}" as it belongs to another skill which haven't written to this memory yet`
      )
    }

    try {
      if (!fs.existsSync(this.memoryPath)) {
        await this.clear()
      }

      return JSON.parse(await fs.promises.readFile(this.memoryPath, 'utf-8'))
    } catch (e) {
      console.error(`Error while reading memory for "${this.name}":`, e)
      throw e
    }
  }

  /**
   * Write the memory
   * @param memory The memory to write
   * @example write({ foo: 'bar' }) // { foo: 'bar' }
   */
  public async write(memory: T): Promise<T> {
    if (!this.isFromAnotherSkill) {
      try {
        await fs.promises.writeFile(
          this.memoryPath,
          JSON.stringify(memory, null, 2)
        )

        return memory
      } catch (e) {
        console.error(`Error while writing memory for "${this.name}":`, e)
        throw e
      }
    } else {
      throw new Error(
        `You cannot write into the memory "${this.name}" as it belongs to another skill`
      )
    }
  }
}

File: bridges/python/src/sdk/memory.py
import json
import os
from typing import TypedDict, Any

from ..constants import SKILL_PATH, SKILLS_PATH


class MemoryOptions(TypedDict, total=False):
    name: str
    default_memory: Any


class Memory:
    def __init__(self, options: MemoryOptions):
        self.name = options['name']
        self.default_memory = options['default_memory'] if 'default_memory' in options else None
        self.memory_path = self.memory_path = os.path.join(
            SKILL_PATH,
            'memory',
            f'{self.name}.json'
        )
        self.__is_from_another_skill = False

        if ':' in self.name and self.name.count(':') == 2:
            self.__is_from_another_skill = True
            domain_name, skill_name, memory_name = self.name.split(':')
            self.memory_path = os.path.join(
                SKILLS_PATH,
                domain_name,
                skill_name,
                'memory',
                memory_name + '.json'
            )

    def clear(self) -> None:
        """
        Clear the memory and set it to the default memory value
        """
        if not self.__is_from_another_skill:
            self.write(self.default_memory)
        else:
            raise ValueError(f'You cannot clear the memory "{self.name}" as it belongs to another skill')

    def read(self):
        """
        Read the memory
        """
        if self.__is_from_another_skill and not os.path.exists(self.memory_path):
            raise ValueError(f'You cannot read the memory "{self.name}" as it belongs to another skill which hasn\'t written to this memory yet')

        try:
            if not os.path.exists(self.memory_path):
                self.clear()

            with open(self.memory_path, 'r') as f:
                return json.load(f)
        except Exception as e:
            print(f'Error while reading memory for "{self.name}": {e}')
            raise e

    def write(self, memory):
        """
        Write the memory
        :param memory: The memory to write
        """
        if not self.__is_from_another_skill:
            try:
                with open(self.memory_path, 'w') as f:
                    json.dump(memory, f, indent=2)

                return memory
            except Exception as e:
                print(f'Error while writing memory for "{self.name}": {e}')
                raise e
        else:
            raise ValueError(f'You cannot write into the memory "{self.name}" as it belongs to another skill')
